### PR TITLE
[PropertyInfo] Fixed promoted property type detection for `PhpStanExtractor`

### LINKED ISSUE #52682: `PhpStanExtractor` doesn't support docblock from promoted properties
### Symfony version(s) affected

6.1+

### Description

Promoted properties may have a doc comment, so the following code (see https://3v4l.org/csbKl):

```php
<?php

class A {
    public function __construct(
        /**
         * @var class-string
         */
        public string $promoted,
    ) {
        // empty
    }
}

$class    = new ReflectionClass(A::class);
$property = $class->getProperty('promoted');

var_dump($property->getDocComment());
```

will dump:

```
string(44) "/**
         * @var class-string
         */"
```

Phpstan itself also reads the type from the promoted property correctly (see https://phpstan.org/r/6de7935b-1d48-49bc-afd2-5748ec5523ac):

image

But the `PhpStanExtractor` ignores the property comment and always tried to use `__construct` comment:

https://github.com/symfony/symfony/blob/37f04cf0ac3fb08ee1ab4a2c1241ec6a926bc696/src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php#L239-L243

### How to reproduce

```php
// First, run "composer require symfony/property-info"
// Then, execute this file:

<?php

use Symfony\Component\PropertyInfo\Extractor\PhpStanExtractor;

require_once __DIR__.'/vendor/autoload.php';

class A {
    public function __construct(
        /**
         * @var class-string
         */
        public string $promoted,
    ) {
        // empty
    }
}

$extractor = new PhpStanExtractor();
$types     = $extractor->getTypes(A::class, 'promoted');

var_dump($types); // NULL
```

### Possible Solution

https://github.com/symfony/symfony/pull/52684

### Additional Context

#46056

### PR DESCRIPTION
**Pull request by LastDragon-ru**
| Q             | A
| ------------- | ---
| Branch?       | 6.3 
| Bug fix?      | yes
| New feature?  | no 
| Deprecations? | no 
| Issues        | Fix #52682 
| License       | MIT

Implementation aligned with actual phpstan behaviour, please see https://phpstan.org/r/8278f037-5a67-4e53-8352-a13585d2eca5

### CODE REVIEW

**Comment by LastDragon-ru** [comment]
Rebased to 6.3

**Comment by LastDragon-ru** [comment]
> fabbot.io — Some changes should be done to comply with our standards.

Looks like it is not related to PR itself :)

**Review by Korbeil — APPROVED** [review]
Seems good to me ! Thanks for the fix :)

File: src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php [hunk 1]
```diff
@@ -239,24 +237,27 @@ private function getDocBlockFromProperty(string $class, string $property): ?arra
             return null;
         }
 
+        // Type can be inside promoted property comment as `@var`
+        $rawDocNode = $reflectionProperty->getDocComment();
+        $phpDocNode = $rawDocNode ? $this->getPhpDocNode($rawDocNode) : null;
         $source = self::PROPERTY;
 
-        if ($reflectionProperty->isPromoted()) {
+        if ($phpDocNode && [] === $phpDocNode->getTagsByName('@var')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php]:
```suggestion
        if (!$phpDocNode?->getTagsByName('@var')) {
```

File: src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php [hunk 2]
```diff
@@ -239,24 +237,27 @@ private function getDocBlockFromProperty(string $class, string $property): ?arra
...
+            $phpDocNode = null;
+        }
+
+        // or in the constructor comment as `@param`.
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php]:
```suggestion
        // or in the constructor as `@param` for promoted properties
```

File: src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php [hunk 3]
```diff
@@ -239,24 +237,27 @@ private function getDocBlockFromProperty(string $class, string $property): ?arra
             return null;
         }
 
+        // Type can be inside promoted property comment as `@var`
```
Comment (nicolas-grekas) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php]:
```suggestion
        // Type can be inside property docblock as `@var`
```

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @LastDragon-ru.

**Comment by LastDragon-ru** [comment]
You are welcome :)

**Pull request merged**